mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -0,0 +1,571 @@
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+
+      if (hasJoinedPartitionColumns) {
+        Schema readDataSchema = TypeUtil.selectNot(readSchema, idColumns);
+        Schema identityPartitionSchema = TypeUtil.select(readSchema, idColumns);
+        return Iterators.transform(
+            open(currentTask, readDataSchema),
+            row -> withPartitionColumns(row, identityPartitionSchema, spec, file.partition()));
+      } else {
+        return open(currentTask, readSchema);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context.getConfiguration());
+      CloseableIterable<T> iterable;
+      switch (file.format()) {
+        case AVRO:
+          iterable = newAvroIterable(inputFile, currentTask, readSchema);
+          break;
+        case ORC:
+          iterable = newOrcIterable(inputFile, currentTask, readSchema);
+          break;
+        case PARQUET:
+          iterable = newParquetIterable(inputFile, currentTask, readSchema);
+          break;
+        default:
+          throw new UnsupportedOperationException(
+              String.format("Cannot read %s file: %s", file.format().name(), file.path()));
+      }
+      currentCloseable = iterable;
+      if (inMemoryDataModel == InMemoryDataModel.DEFAULT) {
+        Expression filter = currentTask.residual();
+        return applyResidualsOnGenericRecords(iterable, filter).iterator();
+      } else {
+        return iterable.iterator();
+      }
+    }
+
+    private Iterable<T> applyResidualsOnGenericRecords(CloseableIterable<T> iterable, Expression filter) {
+      if (filter == null || filter.equals(Expressions.alwaysTrue())) {
+        return iterable;
+      } else {
+        throw new UnsupportedOperationException(
+            String.format("Filter expression %s is not completely satisfied. Additional rows can be returned" +
+                              " not satisfied by the filter expression", filter));
+      }
+    }
+
+    @SuppressWarnings("unchecked")
+    private T withPartitionColumns(T row, Schema identityPartitionSchema, PartitionSpec spec, StructLike partition) {
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement adding partition columns to records for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          return (T) genericRecordWithPartitionsColumns((Record) row, identityPartitionSchema, spec, partition);
+      }
+      return row;
+    }
+
+    private static Record genericRecordWithPartitionsColumns(
+        Record record, Schema identityPartitionSchema, PartitionSpec spec, StructLike partition) {
```

Comment:
There are tests for this in Spark here: https://github.com/apache/incubator-iceberg/commit/6cafdab76864685d5489ba03cd041865b5b57f01

You might consider adding some similar tests.

Reply:
Added similar tests. Had to change the logic of handling identity columns.